### PR TITLE
feat(training): multi-source export with full enrichment JSON for fine-tuning

### PR DESCRIPTION
**Pull request by stevencarpenter**
Export training data from all 5 enrichment sources (shell, agentic sessions, browser, CI workflows, auto-memory) instead of just shell events. Each example includes the source-specific system prompt, a reconstructed user prompt matching the live enrichment input format, and the full validated JSON output the enrichment model produced.

## Changes

- **training.py**: export shell, agentic, browser, workflow, auto-memory nodes with source-appropriate system prompts and full JSON responses. Return value includes per-source counts.
- **__init__.py**: add `hippo-brain export --out DIR --since 30d` subcommand
- **main.rs**: wire `hippo export-training` to delegate to brain export instead of bailing
- **Tests**: 13 pass (8 updated for new format, 3 new: agentic sessions, browser nodes, multi-source coexistence)

## Before

Only shell-event nodes were exported. Assistant responses were just the summary string. One generic system prompt.

## After

All 5 sources produce training pairs with:
- Source-specific system prompts imported from the live enrichment modules
- Reconstructed user prompts matching the enrichment input format
- Full `EnrichmentResult` JSON as the assistant response

Output now includes a `sources` breakdown:
```
Exported 150 examples to ./export/
  train: 120
  valid: 15
  test:  15
  sources:
    agentic: 45
    auto-memory: 12
    browser: 23
    shell: 50
    workflow: 20
```

### CODE REVIEW

**Comment by stevencarpenter** [comment]
@blocks review this pull request

File: brain/src/hippo_brain/training.py [hunk 1]
```diff
@@ -24,83 +154,243 @@ def export_training_data(
...
+            continue
+        try:
+            parsed = json.loads(content)
+        except json.JSONDecodeError, TypeError:
+            continue
+        user_msg = _build_agentic_user_message(sessions)
+        assistant_msg = json.dumps(parsed, ensure_ascii=False)
+        examples.append(
+            {
+                "messages": [
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/training.py:258]: Not a syntax error in this repo: `requires-python = ">=3.14"` and ruff `target-version = "py314"` — PEP 758 (Python 3.14) makes unparenthesized multi-except legal when there is no `as` binding, and ruff at a py314 target actually strips the parentheses. CI is green (module imports fine in the test suite), which it could not be if this were a SyntaxError. No change.

File: brain/src/hippo_brain/training.py [hunk 2]
```diff
@@ -119,4 +409,8 @@ def export_training_data(
         "train": len(train),
         "valid": len(valid),
         "test": len(test),
+        "sources": source_counts,
     }
+
+
+MEMORY_SYSTEM_PROMPT = MEMORY_ENRICHMENT_SYSTEM_PROMPT
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/training.py]: Fixed in 99d620b — the import at the top now aliases directly (`MEMORY_ENRICHMENT_SYSTEM_PROMPT as MEMORY_SYSTEM_PROMPT`) and the trailing alias line is gone.

File: brain/src/hippo_brain/training.py [hunk 3]
```diff
@@ -24,83 +154,243 @@ def export_training_data(
...
+          AND kn.id IN (SELECT DISTINCT knowledge_node_id FROM knowledge_node_browser_events)
+    """
+    browser_params: list = []
+    if since_ms is not None:
+        browser_sql += " AND kn.created_at >= ?"
+        browser_params.append(since_ms)
+
+    for node_id, content in conn.execute(browser_sql, browser_params).fetchall():
+        events = conn.execute(
+            """SELECT be.url, be.title, be.domain, be.dwell_ms, be.scroll_depth, be.search_query
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/training.py:282]: The missing filter is intentional: for workflow nodes the `outcome` column holds the raw GitHub conclusion (success/failure/cancelled/…), not the LLM's success/partial vocab — see the rationale comment in workflow_enrichment.py. Enriched *failed* runs (root cause + one-line fix summaries) are exactly the training signal worth keeping. Documented with a code comment in 99d620b.

File: brain/src/hippo_brain/training.py [hunk 4]
```diff
@@ -24,83 +154,243 @@ def export_training_data(
...
+            parsed = json.loads(content)
+        except json.JSONDecodeError, TypeError:
+            continue
+        user_msg = _build_browser_user_message(events)
+        assistant_msg = json.dumps(parsed, ensure_ascii=False)
+        examples.append(
+            {
+                "messages": [
+                    {"role": "system", "content": BROWSER_SYSTEM_PROMPT},
+                    {"role": "user", "content": user_msg},
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/training.py:302]: Intentional deviation, now documented in the export_training_data docstring (99d620b): min_events counts constituent events and applies to the multi-event sources; workflow and auto-memory nodes are single-input by construction, so applying min_events > 1 would silently exclude those sources entirely rather than filter within them.

File: brain/src/hippo_brain/training.py [hunk 5]
```diff
@@ -24,83 +154,243 @@ def export_training_data(
...
+               JOIN knowledge_node_memory_chunks knmc ON knmc.memory_chunk_id = mc.id
+               WHERE knmc.knowledge_node_id = ?
+               LIMIT 1""",
+            (node_id,),
+        ).fetchone()
+        if rev_row is None:
+            continue
+        revision_text = rev_row[0] or ""
+        if not revision_text.strip():
+            continue
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/training.py:382]: Verified against the live pipeline and this was a real fidelity bug, not just a style gap: enrichment sends build_memory_enrichment_prompt output (repository/path/content-hash header + chunk bodies joined by ---), not raw revision text. Fixed in 99d620b by extracting a shared render_memory_enrichment_input() in auto_memory.py that both the live loop and the exporter use, so exported examples match what the model saw.

File: brain/src/hippo_brain/__init__.py [hunk 6]
```diff
@@ -91,6 +91,85 @@ def _load_runtime_settings() -> dict:
     }
 
 
+def _cmd_serve(args: object) -> None:
+    import uvicorn
+
+    from hippo_brain.server import create_app
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/__init__.py:97]: Keeping the deferred imports deliberately: they isolate each subcommand's dependency cost — `export` doesn't pay for uvicorn/FastAPI/OpenTelemetry, `serve` doesn't pay for the enrichment-module chain, and `--help` pays for neither. All of these are the 'heavy dep' case the deferred-import exception exists for.

File: brain/src/hippo_brain/__init__.py [hunk 7]
```diff
@@ -91,6 +91,85 @@ def _load_runtime_settings() -> dict:
...
+        uvicorn.run(app, host="127.0.0.1", port=settings["port"])
+    finally:
+        if _otel_shutdown:
+            _otel_shutdown()
+
+
+def _cmd_enrich(args: object) -> None:
+    print("Enrichment worker not yet implemented as standalone command.")
+
+
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/src/hippo_brain/__init__.py:138]: Partially adopted in 99d620b: sqlite3.Error (missing/locked/corrupt DB — the realistic operator-facing failures) now prints a clean message and exits 1, and invalid --since exits 1 instead of returning 0. A blanket `except Exception` is deliberately avoided: unexpected errors are bugs and should keep their traceback rather than be masked (this repo treats silent failure as a review blocker).

File: crates/hippo-daemon/src/main.rs [hunk 8]
```diff
@@ -980,10 +980,18 @@ async fn main() -> Result<()> {
...
+            let mut cmd = std::process::Command::new("uv");
+            cmd.args(["run", "--project", "brain", "hippo-brain", "export"])
+                .arg("--out")
+                .arg(&out);
+            if let Some(s) = &since {
+                cmd.arg("--since").arg(s);
+            }
+            let status = cmd.status()?;
+            if !status.success() {
+                anyhow::bail!("hippo-brain export failed (exit {status})");
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply crates/hippo-daemon/src/main.rs:993]: Intentional: `hippo export-training` is an interactive CLI passthrough, and inheriting stdio streams the Python exporter's progress/stats directly to the user. Failure is already surfaced structurally via the exit-status check (`anyhow::bail!` on non-success). If a scripted consumer ever needs structured output it can capture the process output itself; no launchd job invokes this command.

File: brain/tests/test_training.py [hunk 9]
```diff
@@ -72,15 +81,137 @@ def test_export_training_data(tmp_db):
...
+                assert "entities" in assistant
+                assert "tags" in assistant
+                assert "key_decisions" in assistant
 
 
 def test_export_empty_db(tmp_db):
     conn, _ = tmp_db
 
     with tempfile.TemporaryDirectory() as tmpdir:
         stats = export_training_data(conn, tmpdir)
```
[the remark this answers is not in the document]
Comment (stevencarpenter) [reply brain/tests/test_training.py:112]: Fixed in 99d620b — all three tests now assert identity against the imported live prompts (SHELL_SYSTEM_PROMPT, CLAUDE_SYSTEM_PROMPT, BROWSER_SYSTEM_PROMPT) instead of wording substrings.
